### ISSUE TITLE
ADAL in ASP.NET Core Web -> WebApi

### ISSUE THREAD

**Issue by glatzert** [AzureAD/azure-activedirectory-library-for-dotnet#514]
Hey everyone,

I saw the sample od ADAL with an ASP.NET Core Web Application, which uses OpenIdConnect and an Web Api Backend, which uses oAuth Bearer Authentication (https://github.com/Azure-Samples/active-directory-dotnet-webapp-webapi-openidconnect/tree/master/TodoListWebApp/Utils)

Then I implemented the OpenIdConnect-Events, to make sure ADAL get's the Tokens in the Cache. Everything fine so far - but now I have to make sure, my Session is as long accessible as my original authentication cookie. So as tempting as the proposed solution was .. it has some shortcommings.

Looking at the classes involved, we see that we are doing something here, that's already handled by the OpenIdConnect-Implementation of ASP.NET. We resolve the AccessCode ourself with ADAL instead of just using the built-in way.
Additionally we have to do something about the TokenCache, since the default is not web-farm aware.
So we are doing web-farm awareness with session, despite using the existing CookieAutentication Handler of ASP.NET, which does a good job in protecting the Tokens and beeing farm aware.

So I was moving on and thougt something like: "It would be nice to just use OpenIdConnect defaults and not override the events and then just use ADAL to handle AccessToken and RefreshToken "on-demand".
_Bummer_: no `AcquireToken*`-Method takes an AccessToken or another object, which is handled by **me**. 
So I'll have to use the TokenCache...
Next try: Adding a Token to the cache by just calling a `StoreToCache(..)`-Function 
_Bummer_: the method is internal.. (whyyyyyyy?)

Okay then - `class AuthenticationContextTokenCache : TokenCache` will probably work.
My Goal was using the `AuthenticationProperties` from ASP.NETs `AuthenticationContext` to store the current users auth data (which should be easy) and call signIn, in case I have to update the AccessTokens. But again: _bummer_ TokenCache is strictly sync, while AuthenticationContext.GetAuthenticationInfoAsync is strictly async.

Long story short: Is there any usecase, where I can just opt completely out of the TokenCache?
Can you provide a method, which allows me to put in an TokenCacheItem and works on that?
Or can you consider making the TokenCache implement async methods as well?
Currently all approaches seem flawed to me, but perhaps I am missing something..

Thanks for your attention and thanks in advance for any advice.

**Comment by glatzert** [comment]
Would you perhaps consider adding an `ITokenCache`, or an `public void AddCacheItem` or making the `StoreToCache(...)` to an `protected` method? This would enable developers to better use (or I'd say bypass) the TokenCaches in High-Availability-Web-Environments where a single user will always have his own TokenCache instance? (So 1:1 User-to-Request-Mapping)

**Comment by kpanwar** [comment]
you can pass null into the authenticationcontext constructor to bypass the cache. We provide delegates in the token cache for the developer to hook into the cache read/write and persist cache in a DB or some other storage. Check out this article explaining token cache. http://www.cloudidentity.com/blog/2014/07/09/the-new-token-cache-in-adal-v2/

**Comment by glatzert** [comment]
If I pass a null to the AuthContext constructor - how would I present ADAL the AuthToken and RefreshTokens for `AcquireTokenSilent()`?

**Comment by kpanwar** [comment]
You cannot. Other option is to configure delegates to persist cache. You can then create Authcontext and hydrate the cache with the persisted data. See the link i posted about Adal cache.

**Comment by glatzert** [comment]
That is unfortunate. So I'll work around that somehow. Probably I'll just derive from `TokenCache` and use a unhealthy amount of reflection to push my Tokens into the Cache.
Perhaps you would consider an `ITokenCache` or `TokenCache.AddItem`, so we don't have to pull the TokenCache into OpenIdConnectEvents and can just use ADAL for deciding wether a refresh token has to be used or not.

**Comment by kpanwar** [comment]
I am not understanding "my tokens" part. You should be able to supply same token cache instance to auth context created by you and opwnidconnect implementation of asp.net. Or at least provide same delegates for persistence. That will make sure the cache is shared correctly. Does the shared article not solve your issues? Will my suggestion not work either?

**Comment by glatzert** [comment]
The ASP.net OpenIdConnect-Handler will already provide me an ready to use AccessToken and a RefreshToken. The Tokens are also used to determine the lifetime of the Authentication-Session of the user and will be stored to a Cookie (as in Asp.net core samples for oidc) - i am trying not to break that mechanism. - These two Tokens are "my" Tokens (as opposed to ADALs).
Now that AccessToken will expire and the RefreshToken can be used to get new Tokens - 
I thought ADAL would handle that for me in the manner of `AuthContext.AcquireTokenSilentAsync(myAccessToken, myRefreshToken);`, where it would just return the accessToken, if possible or use the refreshToken if neccessary.

I could trade an AccessCode for the AccessToken and RefreshToken in the OIDC-Events of Asp.net by using ADAL and then serialize that into a AuthenticationProperty (so it will be part of the nicely protected auth-cookie). But that is very difficult to archieve, since the TokenCache callbacks are sync and ASP.NETs authentication manager is strictly async.

**Comment by kpanwar** [comment]
Can't a sync method be called asynchronously using task factory? But I do see the problem you have.

**Comment by kpanwar** [comment]
What you propose cannot really be done because if you supply access token from the outside, then Adal has no idea about its expiration time. It will not be able to make that decision for you. I think what you need is calling token cache delegates in an async manner.

**Comment by glatzert** [comment]
Sync mehods can be called async (or just be called), but calling *a*sync methods from a sync method - as serialize and deserialize are - can have difficulties.
Essentially I COULD use `myTask.Result`, or `myTask.Wait()`, but I am not confident, if that'll be a good idea.

Isn't IAT and EXP part of the AccessToken?

I'll take another look into using the TokenCache with an asnyc store and will come back to you - probably tomorrow.

**Comment by glatzert** [comment]
And thanks for your time so far.

**Comment by kpanwar** [comment]
Adal does not look into access tokens and treats them as opaque blob. In fact in the future the tokens could be encrypted as well. Keep me posted!

**Comment by glatzert** [comment]
That just rises another question: if not from the Access Tokens - how does ADAL know it has to use get a new Access token?

**Comment by kpanwar** [comment]
Magic. Just kidding. AAD returns exp_in as a part of the response from the token endpoint. Adal uses it to compute expiration time in UTC and saves it to cache.

**Comment by glatzert** [comment]
Okay - I have a working solution. The only thing I do not like about it, is the fact that it might reauthenticate the user during it's process:

During OpenIdConnect-Events:
`public override async Task AuthorizationCodeReceived(AuthorizationCodeReceivedContext context)
        {
            var clientCred = new ClientCredential(context.Options.ClientId, context.Options.ClientSecret);

            var authContext = new AuthenticationContext(context.Options.Authority, false, new TokenCache());
            await authContext.AcquireTokenByAuthorizationCodeAsync(context.ProtocolMessage.Code,
                new Uri(context.Properties.Items[OpenIdConnectDefaults.RedirectUriForCodePropertiesKey]), clientCred, context.Options.Resource);

            var cachedTokenBytes = authContext.TokenCache.Serialize();
            var cachedToken = Convert.ToBase64String(cachedTokenBytes);

            var properties = new List<AuthenticationToken>
            {
                new AuthenticationToken { Name = TokenNames.TokenCacheProperty, Value = cachedToken }
            };

            context.Properties.StoreTokens(properties);
            context.HandleCodeRedemption();
        }`

That will add the ADAL-Token Cache to the Authentication-Properties. Where we can read it again.
To initialize the Cache (and then use that instance during the whole request) I do this:

`var httpContext = _httpContextAccessor.HttpContext;

            var cachedToken = await httpContext.Authentication.GetTokenAsync(TokenNames.TokenCacheProperty);
            var cachedTokenBytes = Convert.FromBase64String(cachedToken);
            var tokenCache = new TokenCache(cachedTokenBytes);

            return tokenCache;`

So I don't override the original cache implementation, but just feed my serialized cache into that object.
And to repersist the cache after changes, I use this snippet:

`private async Task PersistAdalTokenCache(TokenCache tokenCache)
        {
            if (!tokenCache.HasStateChanged)
                return;

            var httpContext = _httpContextAccessor.HttpContext;

            var newCachedTokenBytes = tokenCache.Serialize();
            var newCachedToken = Convert.ToBase64String(newCachedTokenBytes);

            var authenticateInfo = await httpContext.Authentication.GetAuthenticateInfoAsync(CookieAuthenticationDefaults.AuthenticationScheme);

            var properties = authenticateInfo.Properties;
            properties.Items[TokenNames.TokenCacheProperty] = newCachedToken;

            await httpContext.Authentication.SignInAsync(CookieAuthenticationDefaults.AuthenticationScheme, authenticateInfo.Principal, properties);
        }`

The last snippet will have a "side-effect", since SignInAsync will rewrite the users authentication cookie.

I will probably stick to that solution, since it's quite easy to handle. I have to sort out some minor problems with the TokenCache-Lifetime, but I'll handle that also.

**Comment by kpanwar** [comment]
Looks like you have a working solution. I am closing this issue. Please let me know if you have any questions.
